dbms/src/Storages/Page/V3/PageDirectory.h

```diff
@@ -82,25 +82,134 @@ class PageDirectorySnapshot : public DB::PageStorageSnapshot
+
+    MultiVersionRefCount(const MultiVersionRefCount & other)
+    {
+        if (other.versioned_ref_counts)
+        {
+            versioned_ref_counts = std::make_unique<std::vector<std::pair<PageVersion, Int64>>>();
+            for (const auto & [ver, ref_count] : *(other.versioned_ref_counts))
+            {
+                versioned_ref_counts->emplace_back(ver, ref_count);
+            }
+        }
+    }
+
+    void restoreFrom(const PageVersion & ver, Int64 ref_count)
+    {
+        // versioned_ref_counts being nullptr always means ref count 1
+        RUNTIME_CHECK(ref_count > 0);
+        if (versioned_ref_counts == nullptr && ref_count == 1)
+        {
+            return;
+        }
+        RUNTIME_CHECK(versioned_ref_counts == nullptr);
+        versioned_ref_counts = std::make_unique<std::vector<std::pair<PageVersion, Int64>>>();
+        // empty `versioned_ref_counts` means ref count 1, so the ref count delta here is ref_count - 1
+        versioned_ref_counts->emplace_back(ver, ref_count - 1);
+    }
+
+    void incrRefCount(const PageVersion & ver, Int64 ref_count_delta)
+    {
+        if (versioned_ref_counts == nullptr)
+        {
+            versioned_ref_counts = std::make_unique<std::vector<std::pair<PageVersion, Int64>>>();
+        }
+        versioned_ref_counts->emplace_back(ver, ref_count_delta);
+    }
+
+    void decrRefCountInSnap(UInt64 snap_seq, Int64 deref_count_delta)
+    {
+        if (snap_seq == 0)
+        {
+            // TODO: actually we can collapse versioned_ref_counts here too because there should be no other gc happend concurrently.
+            // But collpase it when `snap_seq` is not zero should be enough. So we just do an append here.
+            versioned_ref_counts->emplace_back(PageVersion(0, 0), -deref_count_delta);
+        }
+        else
+        {
+            auto new_versioned_ref_counts = std::make_unique<std::vector<std::pair<PageVersion, Int64>>>();
+            Int64 ref_count_delta_in_snap = -deref_count_delta;
+            for (const auto & [ver, ref_count_delta] : *versioned_ref_counts)
+            {
+                if (ver.sequence <= snap_seq)
+                {
+                    ref_count_delta_in_snap += ref_count_delta;
+                }
+                else
+                {
+                    new_versioned_ref_counts->emplace_back(ver, ref_count_delta);
+                }
+            }
+            if (ref_count_delta_in_snap == 0 && new_versioned_ref_counts->empty())
+            {
+                versioned_ref_counts = nullptr;
+                return;
+            }
+            RUNTIME_CHECK(ref_count_delta_in_snap > 0);
+            new_versioned_ref_counts->emplace_back(PageVersion(snap_seq, 0), ref_count_delta_in_snap);
+            versioned_ref_counts.swap(new_versioned_ref_counts);
+        }
+    }
+
+    Int64 getRefCountInSnap(UInt64 snap_seq) const
+    {
+        if (versioned_ref_counts == nullptr)
```

Comment:
Minor issue.
```suggestion
        if (!versioned_ref_counts)
        // or if (!versioned_ref_counts == {})
```
So that the compiler don't need to transform `nullptr` to `std:unique_ptr` instance and do comperation.

Reply:
Fixed.